File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.cpp
#include <string>
#include <vector>
#include <stack>
#include <iostream>

using namespace std;

stack<int> st;

int solution(vector<vector<int>> board, vector<int> moves) {
    int answer = 0;
    int n = board.size();
    
    for (auto num : moves) {
        int pick = 0;
        for (int i = 0; i < n; i++) {
            if (board[i][num - 1] != 0){
                pick = board[i][num - 1];
                board[i][num - 1] = 0;
                break;
            }
        }
        
        cout << pick << '\n';
        
        if (pick) {
            if (st.empty()) {
                st.push(pick);
            } else if (st.top() == pick) {
                st.pop();
                answer += 2;
            } else {
                st.push(pick);
            }
        }
    }
    
    return answer;
}

File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.py
from collections import deque

def solution(board, moves):
    answer = 0
    n = len(board)
    dq = deque()
    
    for move in moves:
        pick = 0
        for i in range(n):
            if (board[i][move - 1] != 0):
                pick = board[i][move - 1]
                board[i][move - 1] = 0
                break
                
        if (pick != 0):
            if (len(dq) == 0):
                dq.append(pick)
            elif (dq[-1] == pick):
                dq.pop()
                answer += 2
            else:
                dq.append(pick)
                
    return answer